src/utils.ts

```diff
@@ -0,0 +1,141 @@
+  const weightedCurveSum = (curveA1 + curveA2 + curveA3) / 6n;
+
+  const logCurveA = (floatingCurveA * lnWad((alpha * WAD)
+  / (floatingMaxUtilization - utilizationAfter))) / delta;
+
+  const baseRate = weightedAverageDifference < PRECISION_THRESHOLD ? weightedCurveSum : logCurveA;
+
+  return baseRate + floatingCurveB;
+};
+
+export const marketStateUtilization = (
+  { floatingDebt, floatingAssets }: MarketState,
+): bigint => {
+  if (floatingAssets === 0n) return 0n;
+
+  return (floatingDebt * WAD) / (floatingAssets);
+};
+
+const totalFloatingBorrowAssets = (timestamp: number, marketState: MarketState) => {
+  const {
+    floatingAssets, floatingDebt,
+  } = marketState;
+
+  const utilization = marketState.floatingUtilization;
+  const newUtilization = floatingAssets > 0n
+    ? (floatingDebt * WAD) / floatingAssets
+    : 0n;
+
+  const borrowRate = floatingRate(
+    marketState,
+    min(utilization, newUtilization),
+    max(utilization, newUtilization),
+  );
+
+  return floatingDebt
+  + (floatingDebt * ((borrowRate * BigInt(timestamp - marketState.lastFloatingDebtUpdate))
+  / BigInt(ONE_YEAR_IN_S)))
+  / WAD;
+};
+
+const totalAssets = (timestamp: number, marketState: MarketState) => {
+  const {
+    floatingAssets, floatingDebt, earningsAccumulator,
+    lastAccumulatorAccrual, earningsAccumulatorSmoothFactor,
+  } = marketState;
+
+  const fixedPools = marketState.fixedPools || [];
+  const treasuryFeeRate = marketState.treasuryFeeRate || 0n;
+  const maxFuturePools = fixedPools.length;
+  const elapsed = BigInt(timestamp - lastAccumulatorAccrual);
+  const fixedPoolsAssets = fixedPools.filter(Boolean).reduce((
+    floatingPoolEarnings,
+    { lastAccrual, maturity, unassignedEarnings },
+  ) => (
+    floatingPoolEarnings
+      + (maturity > lastAccrual
+        ? (timestamp < maturity
+          ? (unassignedEarnings * BigInt(timestamp - lastAccrual))
+            / BigInt(maturity - lastAccrual)
+          : unassignedEarnings)
+        : 0n)
+  ), 0n);
+
+  const adjustedEarnings = elapsed && (earningsAccumulator * elapsed)
+      / (elapsed + (earningsAccumulatorSmoothFactor
+        * BigInt(maxFuturePools * FIXED_INTERVAL)) / WAD);
+
+  const netBorrowAssets = ((
+    totalFloatingBorrowAssets(timestamp, marketState)
+    - floatingDebt
+  ) * (WAD - treasuryFeeRate)) / WAD;
+
+  return floatingAssets + fixedPoolsAssets + adjustedEarnings + netBorrowAssets;
+};
+
+const shareValue = (type: 'deposit' | 'borrow', marketState: MarketState, timestamp: number) => {
+  const assets = (type === 'deposit' ? totalAssets : totalFloatingBorrowAssets)(timestamp, marketState);
+  const shares = type === 'deposit' ? marketState.totalSupply : marketState.totalFloatingBorrowShares;
+
+  if (shares === 0n) return 0n;
+  return (assets * WAD) / shares;
+};
+
+export const change = (
+  currentMarketState: MarketState,
+  previousMarketState: MarketState,
```

Comment:
```suggestion
  current: MarketState,
  previous: MarketState,
```

Reply:
why? I'd agree if the function name included 'marketState' in some way, but I think it's unclear what 'current' and 'previous' mean, but let me know if there was another motivation.